primitives/src/policy.rs

```diff
@@ -493,31 +483,31 @@ impl Policy {
 
     /// Returns the supply at a given time (as Unix time) in Lunas (1 NIM = 100,000 Lunas). It is
     /// calculated using the following formula:
-    /// Supply (t) = Genesis_supply + Initial_supply_velocity / Supply_decay * (1 - 2^(- Supply_decay * t))
-    /// Where t is the time in milliseconds since the PoW genesis block and `genesis_supply` is the supply at
+    /// ```text
+    /// supply(t) = total_supply - (total_supply - genesis_supply) * supply_decay^t
+    /// ```
+    /// Where t is the time in milliseconds since the PoS genesis block and `genesis_supply` is the supply at
     /// the genesis of the Nimiq 2.0 chain.
     #[cfg_attr(feature = "ts-types", wasm_bindgen(js_name = supplyAt))]
-    pub fn supply_at(genesis_supply: u64, current_time: u64) -> u64 {
-        assert!(current_time >= Self::POW_GENESIS_TIMESTAMP);
-
-        let t = current_time - Self::POW_GENESIS_TIMESTAMP;
+    pub fn supply_at(genesis_supply: u64, genesis_time: u64, current_time: u64) -> u64 {
+        let t = current_time
+            .checked_sub(genesis_time)
+            .expect("current_time must be greater or equal to genesis_time");
 
-        let supply = genesis_supply
-            + (Self::INITIAL_SUPPLY_VELOCITY / Self::SUPPLY_DECAY
-                * (1.0 - powi(Self::SUPPLY_DECAY_BASE, t))) as u64;
-
-        cmp::min(supply, Policy::TOTAL_SUPPLY)
+        Policy::TOTAL_SUPPLY
+            - ((Policy::TOTAL_SUPPLY - genesis_supply) as f64 * powi(Policy::SUPPLY_DECAY, t))
+                as u64
```

Comment:
Can you explain how the new formula works? It seems you are changing the calculation from a forward calculation (genesis supply + diff) to a backwards calculation (total supply - genesis supply). How does that work, and what is `powi`?

Reply:
Only addressing a tiny part of your comment: `powi` is "to the power of", but implemented in a way that is not different between machines.